# COMP90014 Assignment 1
### Semester 2, 2019

**\*\* This version of the assignment has corrected the guide answer in task 5** 

This assignment should be completed by each student individually. Make sure you read this entire document, and ask for help if anything is not clear. Any changes or clarification to this document will be announced via the LMS.

Please make sure you are aware of the University's rules on academic honesty and plagiarism, which are very strict: https://academichonesty.unimelb.edu.au/ 

Make sure you **do not** copy any code either from other students or from the internet. This is considered plagiarism. It is generally a good idea to avoid looking at any solutions as you may find it surprisingly difficult to generate your own solution to the problem once you have seen somebody else's.

Your completed notebook file containing all your answers will be turned in via LMS. No other files or formats will be accepted - only upload the completed `.ipynb` file.

### Overview
To complete the assignment you will need to finish the tasks in this notebook. There are multiple tasks that are connected in a logical order.

The tasks are a combination of writing your own implementations of algorithms we've discussed in lectures, writing your own code to use library implementations of these algorithms and interpreting the results in short answer format. Each short answer question has a word limit that will be strictly enforced!

In some case, we have provided test input and test output that you can use to try out your solutions. These tests are just samples and are not exhaustive - they may warn you if you've made a mistake, but they are not guaranteed to. It's up to you to decide whether your code is correct.

### Marking

Cells that must be completed to receive marks are clearly labeled. There are 16 graded cells, some of which are code cells, in which you must complete the code to solve a problem, and some of which are markdown cells, in which you must write your answers to short-answer questions. In addition to the graded cells, up to 8 marks will be given for code style, readability, efficiency and comments. 

The total marks for the assignment add up to 45, and it will be worth 15% of your overall subject grade.

### Background and data
During this assignment you will go through the steps to generate trees from a set of sequences. This includes calculating the distance matrices and the trees themselves using multiple algorithms. At each step of the process you will be asked to fill out short answer questions to demonstrate your understanding of the concepts.

The sequences you will use are the following [C2H2 zinc finger genes](https://en.wikipedia.org/wiki/Zinc_finger#Cys2His2) listed with links to their NCBI RefSeq records:
- [*KLF1*](https://www.ncbi.nlm.nih.gov/nuccore/NM_006563)
- [*KLF3*](https://www.ncbi.nlm.nih.gov/nuccore/NM_016531)
- [*KLF7*](https://www.ncbi.nlm.nih.gov/nuccore/NM_003709)
- [*YY1*](https://www.ncbi.nlm.nih.gov/nuccore/NM_003403)
- [*SP1*](https://www.ncbi.nlm.nih.gov/nuccore/NM_001251825)
- [*SP4*](https://www.ncbi.nlm.nih.gov/nuccore/NM_003112)
- [*SP8*](https://www.ncbi.nlm.nih.gov/nuccore/NM_182700)


### Task 1 - Setup

A new directory named `sequences`has been created in the same place as this notebook with a saved  FASTA file for each of the sequences above. The sequence files are named `<accession>.fa` where `<accession>` is the RefSeq accession number (e.g. `NM_006563.fa` for the *KLF1* sequence). Please look in the directory to undertsand the format of the data you will be working with. 

These files should show up when you `git pull`, you can run the code below to read in the sequences you just downloaded into a handy data structure. 

First, we create a dictionary `accessions` with gene names as keys and RefSeq accessions as values and populate it with the seven genes above.

Next we create a second dictionary `sequences` with gene names as keys and DNA sequence objects from the FASTA files as values. We read the FASTA files using the `scikit-bio` library method `skbio.sequence.DNA.read()`.

In [3]:
import os.path
import random
import skbio
from math import inf
from io import StringIO

In [4]:
accessions = dict()
sequences = dict()

accessions['KLF1'] = "NM_006563"
accessions['KLF3'] = "NM_016531"
accessions['KLF7'] = "NM_003709"
accessions['YY1'] = "NM_003403"
accessions['SP1'] = "NM_001251825"
accessions['SP4'] = "NM_003112"
accessions['SP8'] = "NM_182700"

for key in accessions.keys():
    with open(os.path.join("sequences", "{}.fa".format(accessions[key]))) as fasta:
        sequences[key] = skbio.sequence.DNA.read(fasta)      

Run the cell below to see an overview of the information we've just read in. Make sure you understand what is being printed, and ask for help if you don't.

In [5]:
for key in accessions.keys():
    print(key, accessions[key], len(sequences[key]), str(sequences[key])[:10], sep='\t')

KLF1	NM_006563	1645	TCAGAGTTCA
KLF3	NM_016531	5594	AGAGCGAGCG
KLF7	NM_003709	8365	AACAGCGCTG
YY1	NM_003403	6534	CTCCCTCTGC
SP1	NM_001251825	7536	ACCCCCCCCT
SP4	NM_003112	6077	ACAGCCCAGC
SP8	NM_182700	3626	ATTGTATTGC


## Part 1 - Distance calculations
Distances between sequences can be calculated in a number of different ways. In this part, you will calculate distance matrices for a set of sequences in two different ways. 

### Task 2 - Creating a distance matrix
Write a function `init_distance_matrix` that creates a distance matrix as a dictionary of dictionaries (similar to the substitution matrix in the Week 04 - Trees lab). The dictionary keys should be gene names and the values should be `None` after initialization.

In [404]:
# ~~ GRADED CELL (2 marks) - complete this cell ~~

def init_distance_matrix(genes):
    """
    Take in a list of gene names and return a dict.
    The dict should be a nested 'dictionary of dictionaries', where the keys of the
    both the outer dict and the nested dicts are gene names.
    It should be possible to look up a value in the returned dict with a 
    call like dist[gene1][gene2].
    """
    dist = dict()

    ### YOUR CODE HERE - fill in the rest of this function
    for outer_key in genes:
        dist[outer_key] = dict() # create a new nested dist
        for inner_key in genes:
            dist[outer_key][inner_key] = None
        
    return dist

If you have implemented the code correctly, the following tests should give the right results:

In [405]:
# Should give: dict_keys(['KLF1', 'KLF3', 'KLF7', 'YY1', 'SP1', 'SP4', 'SP8'])
dist = init_distance_matrix(accessions.keys())
print(dist.keys())

dict_keys(['KLF1', 'KLF3', 'KLF7', 'YY1', 'SP1', 'SP4', 'SP8'])


In [406]:
# Should give: None
dist = init_distance_matrix(accessions.keys())
print(dist['KLF1']['SP1'])

None


In [407]:
# Should give an error!
dist = init_distance_matrix(accessions.keys())
print(dist['YY1']['SP4']['KLF1'])

TypeError: 'NoneType' object is not subscriptable

The following function will pretty-print a distance matrix that has been specified correctly.

In [365]:
def print_distance_matrix(dist):
    """
    Pretty-print a distance matrix.
    """
    genes = sorted(dist.keys())

    for s in [''] + genes:
        print("{: >10}".format(s), end='')
    print()

    for key1 in genes:
        print("{:10}".format(key1), end='')
        for key2 in genes:
            value = dist[key1][key2]
            if value is None:
                print("      None", end='')
            else:
                print("{: >#10.5f}".format(dist[key1][key2]), end='')
        print()

### Task 3 - Pairwise alignment scores as distances
The `skbio.alignment.local_pairwise_align_ssw()` library function calculates Striped Smith-Waterman alignments using a highly optimized implementation. It returns a tuple containing:
- `TabularMSA` object describing the alignment
- alignment score
- tuple of aligned position ranges

Here is some example output from calling `skbio.alignment.local_pairwise_align_ssw()`. Notice that changing the order of the sequences doesn't change the alignment.

In [11]:
skbio.alignment.local_pairwise_align_ssw(sequences['KLF1'], sequences['YY1'])

(TabularMSA[DNA]
 ----------------------------------
 Stats:
     sequence count: 2
     position count: 34
 ----------------------------------
 CTCCCAGATAAAAAAAAAAAAAAAAAAAAAAAAA
 CTCCCA---AAAAAAAAAAAAAAAAAAAAAAAAA, 53, [(1603, 1636), (4802, 4832)])

In [33]:
skbio.alignment.local_pairwise_align_ssw(sequences['YY1'], sequences['KLF1'])

(TabularMSA[DNA]
 ----------------------------------
 Stats:
     sequence count: 2
     position count: 34
 ----------------------------------
 CTCCCA---AAAAAAAAAAAAAAAAAAAAAAAAA
 CTCCCAGATAAAAAAAAAAAAAAAAAAAAAAAAA, 53, [(4802, 4832), (1603, 1636)])

In [11]:
# Get just the score, by indexing into the tuple
result_tuple = skbio.alignment.local_pairwise_align_ssw(sequences['KLF1'], sequences['YY1'])
result_tuple[1]

53

In [366]:
# Get just the tuples showing aligned ranges
skbio.alignment.local_pairwise_align_ssw(sequences['KLF1'], sequences['YY1'])[2]

[(1603, 1636), (4802, 4832)]

Complete the function `create_pairwise_dist_matrix` below. This function should create a  distance matrix using your function `init_distance_matrix`, then use `skbio.alignment.local_pairwise_align_ssw()` to populate it with alignment scores. `skbio.alignment.local_pairwise_align_ssw()` returns a tuple, and the second item in the tuple is the alignment score. Use `1 / score` as the value in the distance matrix.

Recall that this matrix will be symmetrical so your solution should only calculate each alignment once.

In [408]:
# ~~ GRADED CELL (2 marks) - complete this cell ~~

def create_pairwise_dist_matrix(seqs):
    """
    Take in a dict with sequence names as keys and sequences as values.
    Return a dict representing the pairwise distances between all pairs of sequences.
    """
    ### YOUR CODE HERE - fill in this function
    dist = init_distance_matrix(seqs.keys())
    for outer_idx, outer_key in enumerate(dist):
        for inner_idx, inner_key in enumerate(dist[outer_key]):
            if outer_idx == inner_idx:
                # a gene has 0 distance to itself
                dist[outer_key][inner_key] = 0
            elif outer_idx < inner_idx:
                # generate upper half of the matrix using SSW
                dist[outer_key][inner_key] = \
                1 / skbio.alignment.local_pairwise_align_ssw(seqs[inner_key], seqs[outer_key])[1]
            else:
                # transpose the matrix since it is symmetric
                dist[outer_key][inner_key] = dist[inner_key][outer_key]
                
    return dist

When you have implemented `create_pairwise_dist_matrix` correctly, the following should creat and print the filled-out distance matrix:

In [409]:
dist_pairwise = create_pairwise_dist_matrix(sequences)
print_distance_matrix(dist_pairwise)

                KLF1      KLF3      KLF7       SP1       SP4       SP8       YY1
KLF1         0.00000   0.01220   0.00833   0.02174   0.02500   0.00962   0.01887
KLF3         0.01220   0.00000   0.00855   0.03125   0.02439   0.02941   0.03448
KLF7         0.00833   0.00855   0.00000   0.02222   0.02632   0.02857   0.02326
SP1          0.02174   0.03125   0.02222   0.00000   0.00446   0.00606   0.02564
SP4          0.02500   0.02439   0.02632   0.00446   0.00000   0.01562   0.02326
SP8          0.00962   0.02941   0.02857   0.00606   0.01562   0.00000   0.01852
YY1          0.01887   0.03448   0.02326   0.02564   0.02326   0.01852   0.00000


Test that your function produces the correct results on the toy sequences below. The below code should print

                    GeneA     GeneB     GeneC
    GeneA        0.00000   0.10000   0.25000
    GeneB        0.10000   0.00000   0.25000
    GeneC        0.25000   0.25000   0.00000


In [410]:
test_sequences = {'GeneA': skbio.DNA('GATTACA'), 'GeneB': skbio.DNA('GTTACAT'), 'GeneC':skbio.DNA('AGCTATG')}
test_dist_pairwise = create_pairwise_dist_matrix(test_sequences)
print_distance_matrix(test_dist_pairwise)

               GeneA     GeneB     GeneC
GeneA        0.00000   0.10000   0.25000
GeneB        0.10000   0.00000   0.25000
GeneC        0.25000   0.25000   0.00000


#### Task 3 - Question 1
Why do we use `1 / score` instead of `score` for the distance? (max 50 words).

Write your answer in the markdown cell below.


~~ GRADED CELL (3 marks) - complete this cell ~~

*Your answer here*<br>Smith-Waterman alignment provides a score of alignment, which implies that the higher the score is, the more similar the 2 sequences are. This is inverse to the definition of distance. The larger the distance is, the less similar they are. Hence, we use `1 / score` to keep the definition consistent.

### Task 4 - Multiple alignment distances
The following code will load the precomputed protein multiple sequence alignment of our genes (generated using Clustal $\Omega$) and read it into a `TabularNSA` object. Download the `znfs.clustal_num` file from the LMS and put it in the same directory as the notebook.

In [18]:
with open("znfs.clustal_num") as znfs:
    znf_msa = skbio.TabularMSA.read(znfs, constructor=skbio.Protein)
znf_msa

TabularMSA[Protein]
-----------------------------------------------------------------------
Stats:
    sequence count: 7
    position count: 998
-----------------------------------------------------------------------
--------------------------------- ... ---------------------------------
--------------------------------- ... ---------------------------------
...
MSDQDHSMDEMTAVVKI---------------- ... AICPEGIARLANSGINVMQVADLQSINISGNGF
MSDQKKEEEEEAAAAAAMATEGGKTSEPENNNK ... AISQDSNPATPNVSTNMEE-------------F

Now, you will write some code to calculate pairwise identity from the multiple alignment. The `TabularMSA` is indexed by gene name. We can extract a subset with a pair of genes by indexing with a list of two gene names, using the following syntax, which will be familiar to students that have used `pandas` for data science in Python:

In [19]:
znf_msa.loc[['KLF1', 'YY1']]

TabularMSA[Protein]
-----------------------------------------------------------------------
Stats:
    sequence count: 2
    position count: 998
-----------------------------------------------------------------------
--------------------------------- ... ---------------------------------
--------------------------------- ... ---------------------------------

The following example code iterates over all positions using the `.iter_positons()` method. This returns a `Sequence` object for each column. It skips positions where either sequence contains a gap and counts matches and mismatches. The number of matches and mismatches can be used to calculate the percent identity between two aligned sequences (# matches/ # matches + # mismatches).

In [376]:
match = 0
mismatch = 0
for p1, p2 in znf_msa.loc[['KLF1', 'YY1']].iter_positions():
    if str(p1) == '-' or str(p2) == '-':
        continue
    elif str(p1) == str(p2):
        match += 1
    else:
        mismatch += 1
print("Matches: {}\nMismatches: {}".format(match, mismatch))

Matches: 66
Mismatches: 184


Complete the function `create_msa_dist_matrix` below. This function should start by creating a new distance matrix using your `init_distance_matrix` function from **Task 1**, then fill each cell of the distance matrix with the fraction of non-gap-containing columns with mismatches (or one minus the percent identity) for each pair of sequences in the multiple alignment. As a guide, try generalising the code above.

In [411]:
# ~~ GRADED CELL (2 marks) - complete this cell ~~

def create_msa_dist_matrix(seqs, msa):
    """
    Take in a dict sequences where keys are sequence names and values are sequences, 
    and a multiple sequence alignment object znf_msa.
    Return a dict representing a distance matrix, where the values in the distance
    matrix represent the percent identity between the sequences - i.e. the fraction
    of mismatches out of all matches and mismatches.
    """
    ### YOUR CODE HERE - fill in this function
    dist = init_distance_matrix(seqs.keys())
    for outer_idx, outer_key in enumerate(dist):
        for inner_idx, inner_key in enumerate(dist[outer_key]):
            if outer_idx == inner_idx:
                # a gene has 0 distance to itself
                dist[outer_key][inner_key] = 0
            elif outer_idx < inner_idx:
                # generate upper half of the matrix using MSA
                match = 0
                mismatch = 0
                # reuse the previous code to calculate the numbers of matches and mismatches
                for p1, p2 in msa.loc[[outer_key, inner_key]].iter_positions():
                    if str(p1) == '-' or str(p2) == '-':
                        continue
                    elif str(p1) == str(p2):
                        match += 1
                    else:
                        mismatch += 1
                if match + mismatch != 0:
                    dist[outer_key][inner_key] = mismatch / (match + mismatch)
            else:
                # transpose the matrix since it is symmetric
                dist[outer_key][inner_key] = dist[inner_key][outer_key]
                
    return dist

When you have implemented `create_msa_dist_matrix` correctly, the following will create and print the filled-out percent-identity matrix:

In [412]:
dist_msa = create_msa_dist_matrix(sequences, znf_msa)
print_distance_matrix(dist_msa)

                KLF1      KLF3      KLF7       SP1       SP4       SP8       YY1
KLF1         0.00000   0.64706   0.64552   0.66667   0.69112   0.64655   0.73600
KLF3         0.64706   0.00000   0.61131   0.72245   0.74903   0.68372   0.73745
KLF7         0.64552   0.61131   0.00000   0.66667   0.68398   0.64362   0.78475
SP1          0.66667   0.72245   0.66667   0.00000   0.57642   0.65753   0.78146
SP4          0.69112   0.74903   0.68398   0.57642   0.00000   0.64853   0.76453
SP8          0.64655   0.68372   0.64362   0.65753   0.64853   0.00000   0.76471
YY1          0.73600   0.73745   0.78475   0.78146   0.76453   0.76471   0.00000


Test that your function produces the correct results on the toy sequences below. The below code should print

                   GeneA     GeneB     GeneC
    GeneA        0.00000   0.16667   0.71429
    GeneB        0.16667   0.00000   0.50000
    GeneC        0.71429   0.50000   0.00000

In [413]:
test_sequences = {'GeneA': skbio.DNA('GATTACA'), 'GeneB': skbio.DNA('GTTACAT'), 'GeneC':skbio.DNA('AGCTATG')}
test_multiple_alignment_output = """CLUSTAL O(1.2.4) multiple sequence alignment


GeneA      GATTACA-	7
GeneB      -GTTACAT	7
GeneC      AGCTATG-	7"""
test_msa = skbio.TabularMSA.read(StringIO(test_multiple_alignment_output), constructor=skbio.DNA)
test_msa_dist = create_msa_dist_matrix(test_sequences, test_msa)
print_distance_matrix(test_msa_dist)

               GeneA     GeneB     GeneC
GeneA        0.00000   0.16667   0.71429
GeneB        0.16667   0.00000   0.50000
GeneC        0.71429   0.50000   0.00000


#### Task 4 - Question 1
Do you expect the distance matrices for MSA and SSW to be similar? Why or why not? (max 50 words)

Write your answer in the markdown cell below.

~~ GRADED CELL (3 marks) - complete this cell ~~

*Your answer here*<br>
They are not expected to be similar, both are trying to compare the proportion of the same nucleobase sequences out of the whole though. SSW compares the local distance which attempts to find out the overlaps starting from the different positions. Meanwhile, MSA compares the global distance from the beginning.

### Task 5 - k-mer distances
The k-mer distance between two sequences is the fraction of k-mers that are unique to either sequence. Write a function that calculates a dictionary of k-mers (for `k` = any number) and their counts for each gene sequence. We will not use the counts to calculate the distance but they are useful for other applications.

Be sure to test your function with a short example sequence you can work by hand.

In [414]:
# ~~ GRADED CELL (2 marks) - complete this cell ~~

def count_kmers(seq, k):
    """
    Take in a sequence and a value k.
    Return all kmers of length k in the form of a dict
    with k-mer strings as keys and counts as values.
    """
    ### YOUR CODE HERE - fill in this function
    kmer_dict = dict()
    
    # k must not be larger than the sequence length
    if k <= len(seq):
        n = 0
        while n < len(seq) - k + 1:
            # slide the window of size k
            kmer = seq[n:n+k]
            if kmer in kmer_dict.keys():
                # add 1 to the value of the same existing key
                kmer_dict[kmer] += 1
            else:
                # init a new key
                kmer_dict[kmer] = 1
            
            n += 1
    
    return kmer_dict

When you have implemented `count_kmers` correctly, the following tests should give the right results:

In [415]:
# Should give: {'GTA': 2, 'TAA': 1, 'AAG': 1, 'AGT': 1}
count_kmers('GTAAGTA', 3)

{'GTA': 2, 'TAA': 1, 'AAG': 1, 'AGT': 1}

In [416]:
# Should give: {'GTA': 2, 'TAG': 1, 'AGT': 1, 'TAA': 1}
count_kmers('GTAGTAA', 3)

{'GTA': 2, 'TAG': 1, 'AGT': 1, 'TAA': 1}

In [417]:
# Should give: {'GT': 2, 'TA': 2, 'AG': 1, 'AA': 1}
count_kmers('GTAGTAA', 2)

{'GT': 2, 'TA': 2, 'AG': 1, 'AA': 1}

Write a function `count_kmers_for_sequences` that uses your previous function to count k-mers for all sequences in a dictionary like `sequences`. It should return a new dictionary with gene names as keys and k-mer dictionaries as values.

In [418]:
# ~~ GRADED CELL (2 marks) - complete this cell ~~

def count_kmers_for_sequences(seqs, k):
    """
    Given a dict of sequences, build a dict of the kmer counts for each
    sequence. Keys should be sequence names, and values should be kmer count
    dicts (where keys are kmers and values are counts).
    """
    ### YOUR CODE HERE - fill in this function
    seq_kmer_dict = dict()
    for key in seqs:
        # convert DNA Sequence to String
        seqs_str = skbio.sequence.DNA.__str__(seqs[key])
        # construct a new dict with the kmer numbers
        seq_kmer_dict[key] = count_kmers(seqs_str, k)
    
    return seq_kmer_dict

The below code creates the kmer_counts dict, storing all kmer counts for each sequence. 

It then prints the name of each gene, the number of unique k-mers in each gene and the gene length, one gene per line separated by a tab. If your implementation of `count_kmers_for_sequences` is correct, it should produce the output

**These numbers have been corrected** 
    
    KLF1	1111	1645
    KLF3	2633	5594
    KLF7	3044	8365
    YY1	 2858	6534
    SP1	 2756	7536
    SP4	 2522	6077
    SP8	 2012	3626


In [419]:
kmer_counts = count_kmers_for_sequences(sequences, k=6)

for key in kmer_counts.keys():
    print(key, len(kmer_counts[key]), len(sequences[key]), sep='\t')

KLF1	1111	1645
KLF3	2633	5594
KLF7	3044	8365
YY1	2858	6534
SP1	2756	7536
SP4	2522	6077
SP8	2012	3626


#### Task 5 - Question 1
How well correlated are the unique k-mer counts and the gene lengths? What does this suggest about the sequences? (max 50 words)

Write your answer in the markdown cell below.

~~ GRADED CELL (3 marks) - complete this cell ~~

*Your answer here*<br>
In general, the longer the gene is, the more unique k-mers it has (positive correlation). However, this is not always the case since a sequence may have a number of repeated segments whose k-mers are not unique. This implies the sequence becomes more and more unique as it goes longer.

Now write a function to calculate the k-mer distance between a pair of genes given their k-mer count dictionaries. Recall that the k-mer distance is the fraction of k-mers that are unique to one of the sequences. You may find the Python `set` structure useful.

In [420]:
# ~~ GRADED CELL (2 marks) - complete this cell ~~

def kmer_distance(k1, k2):
    """
    Given two k-mer count dictionaries, calculate the k-mer distance between these two sets of kmers.
    The counts themselves are not used. Return a floating-point number for the distance.
    """
    ### YOUR CODE HERE - fill in this function
    total_kmer = 0
    identical_kmer = 0
    dist = None
    
    # counts dict must not be both empty
    if len(k1) > 0 or len(k2) > 0:
        # simply count the total kmers in 2 dicts
        total_kmer = len(k1) + len(k2)

        for kmer_1 in k1:
            for kmer_2 in k2:
                if kmer_1 == kmer_2:
                    identical_kmer += 1

        # calculate all the unique kmers
        unique_kmer = total_kmer - identical_kmer
        # "unique_kmer - identical_kmer" are the counts of the distinct kmers
        dist = (unique_kmer - identical_kmer) / unique_kmer
    
    return dist

When you have implemented the `kmer_distance` function correctly, this test should give the right result:

In [421]:
# Should give: 0.4
test_counts_1 = {'GTA': 2, 'TAA': 1, 'AAG': 1, 'AGT': 1} 
test_counts_2 = {'GTA': 2, 'TAG': 1, 'AGT': 1, 'TAA': 1} 
kmer_distance(test_counts_1, test_counts_2)

0.4

Complete the function `create_kmer_dist_matrix` below. This function should create a new distance matrix containining k-mer distances. The matrix, as before, should be returned as a dict of dicts. Use your `kmer_distance` function and your `init_distance_matrix` function.

In [422]:
# ~~ GRADED CELL (2 marks) - complete this cell ~~

def create_kmer_dist_matrix(counts):
    """
    Take in a kmer counts dict, where keys are gene names and values are
    dicts of kmer counts (i.e. kmers as keys and counts as values). 
    Return a kmer distance matrix in the form of a dict of dicts of distances.
    """
    ### YOUR CODE HERE - fill in this function
    dist = init_distance_matrix(counts.keys())
    for outer_idx, outer_key in enumerate(dist):
        for inner_idx, inner_key in enumerate(dist[outer_key]):
            if outer_idx == inner_idx:
                # a gene has 0 distance to itself
                dist[outer_key][inner_key] = 0
            elif outer_idx < inner_idx:
                # calculate the upper half of the dict using self-coded k-mer distance
                dist[outer_key][inner_key] = kmer_distance(counts[outer_key], counts[inner_key])
            else:
                # transpose the matrix since it is symmetric
                dist[outer_key][inner_key] = dist[inner_key][outer_key]
                
    return dist

In [423]:
dist_kmer = create_kmer_dist_matrix(kmer_counts)
print_distance_matrix(dist_kmer)

                KLF1      KLF3      KLF7       SP1       SP4       SP8       YY1
KLF1         0.00000   0.76517   0.74585   0.75659   0.78169   0.72687   0.73073
KLF3         0.76517   0.00000   0.39087   0.36845   0.41385   0.57951   0.40656
KLF7         0.74585   0.39087   0.00000   0.32997   0.38056   0.53661   0.36010
SP1          0.75659   0.36845   0.32997   0.00000   0.36594   0.56602   0.36707
SP4          0.78169   0.41385   0.38056   0.36594   0.00000   0.60191   0.41858
SP8          0.72687   0.57951   0.53661   0.56602   0.60191   0.00000   0.54148
YY1          0.73073   0.40656   0.36010   0.36707   0.41858   0.54148   0.00000


If you've implemented `create_kmer_dist_matrix` correctly, the following test case should display

                   GeneA     GeneB
    GeneA        0.00000   0.40000
    GeneB        0.40000   0.00000

In [424]:
test_counts = {'GeneA': {'GTA': 2, 'TAA': 1, 'AAG': 1, 'AGT': 1},
               'GeneB': {'GTA': 2, 'TAG': 1, 'AGT': 1, 'TAA': 1}} 
test_kmer_dist_matrix = create_kmer_dist_matrix(test_counts)
print_distance_matrix(test_kmer_dist_matrix)

               GeneA     GeneB
GeneA        0.00000   0.40000
GeneB        0.40000   0.00000


Check your implementation by completing the `create_kmer_dist_matrix_skbio` function below. It is very similar to the function above, but in this case you build a distance matrix by calculating k-mer distance using `skbio.sequence.distance.kmer_distance()`. The library function takes as input two sequences (from your `sequences` dictionary) and a value for `k`.

In [425]:
# ~~ GRADED CELL (2 marks) - complete this cell ~~

def create_kmer_dist_matrix_skbio(seqs):
    """
    Take in a kmer counts dict, where keys are gene names and values are
    dicts of kmer counts (i.e. kmers as keys and counts as values). 
    Return a kmer distance matrix in the form of a dict of dicts of distances.
    
    Use skbio.sequence.distance.kmer_distance() to calculate distances for each cell.
    """
    ### YOUR CODE HERE - fill in this function
    k = 6
    dist = init_distance_matrix(seqs.keys())
    for outer_idx, outer_key in enumerate(dist):
        for inner_idx, inner_key in enumerate(dist[outer_key]):
            if outer_idx == inner_idx:
                # a gene has 0 distance to itself
                dist[outer_key][inner_key] = 0
            elif outer_idx < inner_idx:
                # calculate the upper half of the dict using skbio k-mer distance
                dist[outer_key][inner_key] = skbio.sequence.distance.kmer_distance(seqs[outer_key], seqs[inner_key], k)
            else:
                # transpose the matrix since it is symmetric
                dist[outer_key][inner_key] = dist[inner_key][outer_key]
                
    return dist

In [426]:
dist_kmer_lib = create_kmer_dist_matrix_skbio(sequences)
print_distance_matrix(dist_kmer_lib)

                KLF1      KLF3      KLF7       SP1       SP4       SP8       YY1
KLF1         0.00000   0.76517   0.74585   0.75659   0.78169   0.72687   0.73073
KLF3         0.76517   0.00000   0.39087   0.36845   0.41385   0.57951   0.40656
KLF7         0.74585   0.39087   0.00000   0.32997   0.38056   0.53661   0.36010
SP1          0.75659   0.36845   0.32997   0.00000   0.36594   0.56602   0.36707
SP4          0.78169   0.41385   0.38056   0.36594   0.00000   0.60191   0.41858
SP8          0.72687   0.57951   0.53661   0.56602   0.60191   0.00000   0.54148
YY1          0.73073   0.40656   0.36010   0.36707   0.41858   0.54148   0.00000


### Task 6 - Building trees with UPGMA
In this section you will implement the UPGMA algorithm to create ultrameric trees as described in lectures. You will output the trees in Newick format and import those Newick strings into an `skbio` library function to draw the trees as ASCII art.

To implement UPGMA, you should split the process into computable steps, write a function for each step and then use the functions to build your tree.

The first step of the UPGMA algorithm is finding the minimum distance in the matrix. Write a function `minimum_distance` that takes a distance matrix and returns a tuple with the distance and the two node names corresponding to the minimum distance.

In [427]:
# ~~ GRADED CELL (2 marks) - complete this cell ~~

def minimum_distance(dist):
    """
    Find and return a tuple containing the minimum distance
    in the distance matrix and the nodes names corresponding 
    to the minimum.
    """
    ### YOUR CODE HERE - fill in this function
    min_val = None
    gene_1 = None
    gene_2 = None
    
    # the dict size must be larger than 1
    if len(dist) > 1:
        # construct an iteration of the input dict
        dist_iter = iter(dist)
        # get the first 2 keys of the dict
        gene_1 = next(dist_iter)
        gene_2 = next(dist_iter)
        # assume the nodes with the first 2 keys have the minimum distance
        min_val = dist[gene_1][gene_2]

        for outer_idx, outer_key in enumerate(dist):
            for inner_idx, inner_key in enumerate(dist[outer_key]):
                # we just need to compare the upper half of the dict since it is symmetric
                if outer_idx < inner_idx:
                    if dist[outer_key][inner_key] < min_val:
                        # assign min_val and gene names if there is a value smaller than the previous minimum
                        min_val = dist[outer_key][inner_key]
                        gene_1 = outer_key
                        gene_2 = inner_key

    return min_val, gene_1, gene_2

When you have implemented `minimum_distance` correctly, the following tests should give the right result:

In [428]:
# Should give: (3.3, 'geneB', 'geneC')
test_dist = {'geneA': {'geneA':0, 'geneB':5, 'geneC':12},
             'geneB': {'geneA':5, 'geneB':0, 'geneC':3.3},
             'geneC': {'geneA':12, 'geneB':3.3, 'geneC':0}}
print(minimum_distance(test_dist))

(3.3, 'geneB', 'geneC')


In [429]:
# Should give: (4.1, 'geneA', 'geneB')
test_dist = {'geneA': {'geneA':0, 'geneB':4.1, 'geneC':5},
             'geneB': {'geneA':4.1, 'geneB':0, 'geneC':7.2},
             'geneC': {'geneA':5, 'geneB':7.2, 'geneC':0}}
print(minimum_distance(test_dist))

(4.1, 'geneA', 'geneB')


Write another function `update_distances` that takes a distance matrix, the names of the nodes being combined, the name of the newly created node and a dictionary of node sizes and returns a new distance matrix with the result of merging and creating the new node.

In [430]:
# ~~ GRADED CELL (2 marks) - complete this cell ~~

def update_distances(dist, a, b, new_node, node_sizes):
    """
    Take in:
      dist : a distance matrix, structured as a dict of dicts of distances,
             with node names as keys
      a, b : names of nodes to merge
      new_node : the name of the new node to create 
      node_sizes : a dict giving the weightings (cluster sizes) of existing nodes,
                   with names as keys and sizes as values
    Return the new distance matrix, which will contain a node new_node
    and will not contain a or b.
    """
    ### YOUR CODE HERE - fill in this function
    # hard copy the dist dict since we want to modify the keys
    dist_copy = dict.copy(dist)
    # remove the 2 keys to be combined
    dist_copy.pop(a, None)
    dist_copy.pop(b, None)
    # append the new key to be created and init the value to be None
    dist_copy[new_node] = None

    # init the new dict using the keys of the copied dict
    new_dist = init_distance_matrix(dist_copy.keys())
    for outer_idx, outer_key in enumerate(new_dist):
        for inner_idx, inner_key in enumerate(new_dist[outer_key]):
            if outer_idx == inner_idx:
                # a gene has 0 distance to itself
                new_dist[outer_key][inner_key] = 0
            elif outer_idx < inner_idx:
                ori_key = ''
                if inner_key not in dist.keys():
                    # the new key is inner_key (column) so the original key ori_key is outer_key
                    ori_key = outer_key
                    # calculate the new distance based on the weight of cluster sizes
                    new_dist[outer_key][inner_key] = \
                    (dist[ori_key][a] * node_sizes[a] + dist[ori_key][b] * node_sizes[b]) / (node_sizes[a] + node_sizes[b])
                elif outer_key not in dist.keys():
                    # the new key is outer_key (row) so the original key ori_key is inner_key
                    ori_key = inner_key
                    # calculate the new distance based on the weight of cluster sizes
                    new_dist[outer_key][inner_key] = \
                    (dist[ori_key][a] * node_sizes[a] + dist[ori_key][b] * node_sizes[b]) / (node_sizes[a] + node_sizes[b])
                else:
                    # there is no new key to be calculated so we just copy the value
                    new_dist[outer_key][inner_key] = dist[outer_key][inner_key]
            else:
                # transpose the matrix since it is symmetric
                new_dist[outer_key][inner_key] = new_dist[inner_key][outer_key]
            
    return new_dist

Here are some test cases you can use in testing `update_distances`:

The below cell should print 

                   geneC    nodeAB
    geneC        0.00000   6.10000
    nodeAB       6.10000   0.00000

In [431]:
# Should return {'geneC': {'geneC': 0, 'nodeAB': 6.1}, 'nodeAB': {'geneC': 6.1, 'nodeAB': 0.0}}, 
# and also print the test output as shown above
test_dist = {'geneA': {'geneA':0, 'geneB':4.1, 'geneC':5},
             'geneB': {'geneA':4.1, 'geneB':0, 'geneC':7.2},
             'geneC': {'geneA':5, 'geneB':7.2, 'geneC':0}}
test_sizes = {'geneA': 1, 'geneB': 1, 'geneC': 1}
test_output = update_distances(test_dist, 'geneA', 'geneB', 'nodeAB', test_sizes)
print(test_output)
print_distance_matrix(test_output)

{'geneC': {'geneC': 0, 'nodeAB': 6.1}, 'nodeAB': {'geneC': 6.1, 'nodeAB': 0}}
               geneC    nodeAB
geneC        0.00000   6.10000
nodeAB       6.10000   0.00000


The below cell should print 

                   geneC    nodeAB
    geneC        0.00000   5.55000
    nodeAB       5.55000   0.00000

In [432]:
# Should return {'geneC': {'geneC': 0, 'nodeAB': 5.55}, 'nodeAB': {'geneC': 5.55, 'nodeAB': 0.0}}, 
# and also print the test output as shown above
test_dist = {'nodeA': {'nodeA':0, 'geneB':4.1, 'geneC':5},
             'geneB': {'nodeA':4.1, 'geneB':0, 'geneC':7.2},
             'geneC': {'nodeA':5, 'geneB':7.2, 'geneC':0}}
test_sizes = {'nodeA': 3, 'geneB': 1, 'geneC': 1}
test_output = update_distances(test_dist, 'nodeA', 'geneB', 'nodeAB', test_sizes)
print(test_output)
print_distance_matrix(test_output)

{'geneC': {'geneC': 0, 'nodeAB': 5.55}, 'nodeAB': {'geneC': 5.55, 'nodeAB': 0}}
               geneC    nodeAB
geneC        0.00000   5.55000
nodeAB       5.55000   0.00000


Write a final function `upgma` that takes a distance matrix as input and returns a tree as a Newick formatted string.

Because we want to output our tree in Newick format, you will need an additional dictionary of Newick strings for each node that gets created. You also need to keep track of the branch lengths. Recall that the $D_{ij}$ in the algorithm from lecture is the length to the tip not the internal branch length, but the Newick string expects the internal branch lengths.

In [452]:
# ~~ GRADED CELL (2 marks) - complete this cell ~~

def upgma(dist):
    """
    Build a UPGMA tree based on the given distance matrix
    and return the tree as a string in Newick format.
    """
    ### YOUR CODE HERE - fill in this function
    # init final result string
    upgma_str = ''
    # init cluster size dict
    node_sizes = dict()
    # init internal branch length dict
    node_int_dist = dict()
    # init external branch length dict
    node_ext_dist = dict()
    # init Newick string dict to record the result string of each iteration
    newick_str_dict = dict()
    # init a reversed newick_str_dict to assemble the final result
    rev_newick_str_dict = dict()
    
    # init the values in node_sizes to be all 1
    for key in dist:
        node_sizes[key] = 1
    
    # assign dist to be the dict of the first iteration
    curr_dist = dist
    i = 0
    while i < len(dist) - 1:
        # find the minimum distance and its nodes using the previous function
        val, gene_1, gene_2 = minimum_distance(curr_dist)
        
        # for both nodes
        for gene in [gene_1, gene_2]:
            if node_sizes[gene] == 1:
                # internal length is a half of the distance if the node has not been combined
                node_int_dist[gene] = val / 2
            else:
                # init a filtered_dict to construct a dict that only contains its preceding nodes
                filtered_dist = dict()
                for dist_key in node_ext_dist:
                    if dist_key in gene:
                        filtered_dist[dist_key] = node_ext_dist[dist_key]
                
                # find the maximum 
                max_key = max(filtered_dist, key = filtered_dist.get)
                """internal length of a newly combined node is 
                a half of its distance subtract the maximum external length of 
                its preceding nodes"""
                node_int_dist[gene] = val / 2 - filtered_dist[max_key]
                
            # calculate the external length
            node_ext_dist[gene] = val / 2
            
        # update current dist for next iteration
        curr_dist = update_distances(curr_dist, gene_1, gene_2, gene_1 + gene_2, node_sizes)
        # update the weights and keys of node_sizes dict
        node_sizes[gene_1 + gene_2] = node_sizes[gene_1] + node_sizes[gene_2]
        node_sizes.pop(gene_1, None)
        node_sizes.pop(gene_2, None)
        
        # format the Newick string of each iteration and store it in a dict
        newick_str_dict[gene_1 + gene_2] = \
    "({}:{},{}:{})".format(gene_1, node_int_dist[gene_1], gene_2, node_int_dist[gene_2])
            
        i += 1
    
    # reverse newick_str_dict enabling the assembly outside-in
    rev_newick_str_dict_keylist = list(newick_str_dict.keys())[::-1]
    for key in rev_newick_str_dict_keylist:
        rev_newick_str_dict[key] = newick_str_dict[key]
    
    # reversedly loop the newick_str_dict
    for idx, key in enumerate(rev_newick_str_dict):
        if idx == 0:
            # assign the first Newick string to be the result from the last iteration
            upgma_str = newick_str_dict[key]
        else:
            # always replace the key with the value from the previous combination of the same key
            upgma_str = upgma_str.replace(key, newick_str_dict[key])
            
    # append a ";" to the Newick string to finalise it
    upgma_str += ";"
    
    return upgma_str

Here is a test case you can use to help you test your `upgma` solution:

In [453]:
# Should give: '(geneC:3.05,(geneA:2.05,geneB:2.05):1.0);'
test_dist = {'geneA': {'geneA':0, 'geneB':4.1, 'geneC':5},
             'geneB': {'geneA':4.1, 'geneB':0, 'geneC':7.2},
             'geneC': {'geneA':5, 'geneB':7.2, 'geneC':0}}
upgma(test_dist)

'(geneC:3.05,(geneA:2.05,geneB:2.05):1.0);'

Let's try UPGMA out using different distance measures. The following three blocks generate trees using your `upgma` function and print those trees to the console. There is one block per distance method.

In [454]:
newick = upgma(dist_pairwise)
t = skbio.io.read(StringIO(newick), format="newick", into=skbio.tree.TreeNode)
print("Local alignment score distance, UPGMA")
print()
print(newick)
print()
print(t.ascii_art())

Local alignment score distance, UPGMA

((KLF3:0.005185532624557015,(KLF1:0.004166666666666667,KLF7:0.004166666666666667):0.0010188659578903485):0.007111236584955631,(YY1:0.011235893018838755,(SP8:0.005421401515151515,(SP1:0.002232142857142857,SP4:0.002232142857142857):0.0031892586580086583):0.00581449150368724):0.0010608761906738904);

                    /-KLF3
          /--------|
         |         |          /-KLF1
         |          \--------|
---------|                    \-KLF7
         |
         |          /-YY1
          \--------|
                   |          /-SP8
                    \--------|
                             |          /-SP1
                              \--------|
                                        \-SP4


In [455]:
newick = upgma(dist_msa)
t = skbio.io.read(StringIO(newick), format="newick", into=skbio.tree.TreeNode)
print("Multiple sequence alignment percent identity distance, UPGMA")
print()
print(t)
print()
print(t.ascii_art())

Multiple sequence alignment percent identity distance, UPGMA

(YY1:0.38074116087996285,((KLF1:0.32314530289727833,(KLF3:0.30565371024734983,KLF7:0.30565371024734983):0.0174915926499285):0.018732981142131566,(SP8:0.326515080918212,(SP1:0.28820960698689957,SP4:0.28820960698689957):0.03830547393131245):0.015363203121197877):0.03886287684055295);


          /-YY1
         |
         |                    /-KLF1
---------|          /--------|
         |         |         |          /-KLF3
         |         |          \--------|
          \--------|                    \-KLF7
                   |
                   |          /-SP8
                    \--------|
                             |          /-SP1
                              \--------|
                                        \-SP4


In [456]:
newick = upgma(dist_kmer)
t = skbio.io.read(StringIO(newick), format="newick", into=skbio.tree.TreeNode)
print("K-mer distance, UPGMA")
print()
print(t)
print()
print(t.ascii_art())

K-mer distance, UPGMA

(KLF1:0.3755750301864033,(SP8:0.28255306393592583,(KLF3:0.19746630247530178,(SP4:0.19418100427492976,(YY1:0.18179347698709536,(KLF7:0.16498704290238986,SP1:0.16498704290238986):0.0168064340847055):0.012387527287834404):0.0032852982003720166):0.08508676146062405):0.09302196625047748);


          /-KLF1
---------|
         |          /-SP8
          \--------|
                   |          /-KLF3
                    \--------|
                             |          /-SP4
                              \--------|
                                       |          /-YY1
                                        \--------|
                                                 |          /-KLF7
                                                  \--------|
                                                            \-SP1


#### Task 6 - Question 1
Why are the trees different? (max 50 words)

Write your answer in the markdown cell below.

~~ GRADED CELL (3 marks) - complete this cell ~~

*Your answer here*<br>
Because the Newick strings are distinct. The strings are generated with UPGMA algorithm according to the input distance matrix, calculated based on different algorithms, resulting in different distance values. Since UPGMA searches for the nodes with minimum distance and combines them iteratively, different minimum values may result in distinct combinations.

### Task 7 - Building trees with neighbor joining
In this section you will use the `skbio` implementation of the neighbor joining algorithm to build trees and visualize them as in the previous section.

The following code converts a dictionary-based distance matrix to a `DistanceMatrix` object suitable for the `skbio` neighbor joining implementation.

In [438]:
def convert_distance_matrix(dist):
    genes = sorted(dist.keys())
    dist_lists = list()

    for key1 in genes:
        a = list()
        for key2 in genes:
            a.append(dist[key1][key2])
        dist_lists.append(a)
        
    return skbio.DistanceMatrix(dist_lists, genes)

The following three blocks generate trees using the `skbio` neighbor joining implementation and print those trees to the console. There is one block per distance method.

In [439]:
m = convert_distance_matrix(dist_pairwise)
t = skbio.tree.nj(m)
print("Local alignment score distance, NJ")
print()
print(t)
print()
print(t.ascii_art())

Local alignment score distance, NJ

((SP4:0.003585,SP1:0.000879):0.006023,(YY1:0.011673,(KLF1:0.000791,(KLF7:0.00197,KLF3:0.006577):0.0052):0.007063):0.004389,SP8:0.002588);


                    /-SP4
          /--------|
         |          \-SP1
         |
         |          /-YY1
         |---------|
---------|         |          /-KLF1
         |          \--------|
         |                   |          /-KLF7
         |                    \--------|
         |                              \-KLF3
         |
          \-SP8


In [440]:
m = convert_distance_matrix(dist_msa)
t = skbio.tree.nj(m)
print("Multiple sequence alignment percent identity distance, NJ")
print()
print(t)
print()
print(t.ascii_art())

Multiple sequence alignment percent identity distance, NJ

((SP8:0.312767,(SP4:0.292451,SP1:0.283968):0.052054):0.020465,(KLF1:0.3114,(KLF7:0.298809,KLF3:0.312498):0.029237):0.003537,YY1:0.421869);


                    /-SP8
          /--------|
         |         |          /-SP4
         |          \--------|
         |                    \-SP1
         |
---------|          /-KLF1
         |---------|
         |         |          /-KLF7
         |          \--------|
         |                    \-KLF3
         |
          \-YY1


In [441]:
m = convert_distance_matrix(dist_kmer)
t = skbio.tree.nj(m)
print("K-mer distance, NJ")
print()
print(t)
print()
print(t.ascii_art())

K-mer distance, NJ

(SP1:0.160175,(KLF3:0.205666,(KLF7:0.165764,(YY1:0.178757,(SP8:0.267982,KLF1:0.458883):0.093916):0.016849):0.009163):0.002511,SP4:0.205769);


          /-SP1
         |
         |          /-KLF3
         |---------|
         |         |          /-KLF7
         |          \--------|
---------|                   |          /-YY1
         |                    \--------|
         |                             |          /-SP8
         |                              \--------|
         |                                        \-KLF1
         |
          \-SP4


#### Task 7 - Question 1
Based on all the trees you've generated, which distance method do you think is the best and why? (max 100 words)

Write your answer in the markdown cell below.

~~ GRADED CELL (3 marks) - complete this cell ~~

*Your answer here*<br>
Multiple sequence alignment (MSA) method is the best. From the algorithm perspective, both Striped Smith-Waterman (SSW) and K-mer focus on the similarity/distance of segmented sequences. In comparison, MSA considers the entire distance values as a whole, which is more used in homology analysis. From the trees, all the gene sequences with same prefix (KLF, YY, SP) are categorised into the same branches (same family), no matter which algorithm we use (UPGMA or NJ). This turns out that the accuracy of MSA is the highest among the three, indicating that sequences with similar names are more likely to be homologous.